js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -686,94 +243,153 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
-                    continue;
+    private constructWithRelatedRecords(
+        key: ViewModelFieldPaths<ViewModelClassType>,
+        baseRecord: PartialViewModel<ViewModelClassType>
+    ): PartialViewModel<ViewModelClassType> | null {
+        const newData = { ...baseRecord._data } as Record<string, any>;
+        for (const [relationFieldName, relationFieldPath] of Object.entries(key.relations)) {
+            const relationField = this.viewModel.getField(
+                relationFieldName
+            ) as BaseRelatedViewModelField<any, any, any>;
+            const idOrIds = baseRecord[relationField.sourceFieldName];
+            if (idOrIds != null) {
+                if (Array.isArray(idOrIds)) {
+                    const records = relationField.cache.getList(idOrIds, relationFieldPath, true);
+                    if (records.length !== idOrIds.length) {
+                        return null;
+                    }
+                    newData[relationFieldName] = records;
+                } else {
+                    const record = relationField.cache.get(idOrIds, relationFieldPath);
+                    if (!record) {
+                        return null;
+                    }
+                    newData[relationFieldName] = record;
                 }
-                const newRecord = this.createRecordForSubKey(
-                    record,
-                    fieldsKey,
-                    this.reverseCacheKey(key)
+            } else {
+                newData[relationFieldName] = null;
+            }
+        }
+        return new this.viewModel(newData) as PartialViewModel<ViewModelClassType>;
+    }
+
+    private relationListenerPaths = new Map<
+        PartialViewModel<ViewModelClassType>,
+        ViewModelFieldPaths<ViewModelClassType>[]
+    >();
+
+    /**
+     * When a listener is added to a record for a set of fields we need to also listen
+     * to changes on the caches for any related records. This functions handles setting
+     * up those listeners.
+     * @param key The key the listener was added to. The relations that need to be listened to are extracted from this.
+     * @param record The last record that the listener would have been notified about. When the listener is first
+     * added this will be the current record in the cache. When the record changes this will be the new record that
+     * was just added.
+     * @private
+     */
+    private setupRelationListeners(
+        key: ViewModelFieldPaths<ViewModelClassType>,
+        record: PartialViewModel<ViewModelClassType>
+    ): void {
+        let registeredPaths = this.relationListenerPaths.get(record);
+        if (!registeredPaths) {
+            registeredPaths = [];
+            this.relationListenerPaths.set(record, registeredPaths);
+        }
+        if (registeredPaths.includes(key)) {
+            return;
+        }
+        registeredPaths.push(key);
+        this.relationListenerUnsubscribe.get(key)?.forEach(unsub => unsub());
```

Comment:
I'll add this comment

```
        // This may already be set for a previous call to this function for a different record (ie.
        // a previous version of the record with different related record id(s)). In that case remove
        // the old listeners and add new ones to ensure they match the latest related record id(s).
```